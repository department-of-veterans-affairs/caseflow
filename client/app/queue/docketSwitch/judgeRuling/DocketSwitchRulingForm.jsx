import React, { useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';

import { useForm, Controller } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';

import {
  DOCKET_SWITCH_RULING_TITLE,
  DOCKET_SWITCH_RULING_INSTRUCTIONS,
} from 'app/../COPY';
import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';
import Button from 'app/components/Button';
import { sprintf } from 'sprintf-js';
import DISPOSITIONS from 'constants/DOCKET_SWITCH_DISPOSITIONS';
import TextareaField from 'app/components/TextareaField';
import RadioField from 'app/components/RadioField';
import SearchableDropdown from 'app/components/SearchableDropdown';
import ReactMarkdown from 'react-markdown';
import gfm from 'remark-gfm';
import { css } from 'glamor';

const schema = yup.object().shape({
  disposition: yup.
    mixed().
    oneOf(Object.keys(DISPOSITIONS)).
    required(),
  context: yup.string(),
  attorney: yup.
    object().
    shape({ label: yup.string().required(), value: yup.number().required() }).
    required(),
});

export const DocketSwitchRulingForm = ({
  appellantName,
  defaultAttorneyId,
  clerkOfTheBoardAttorneys = [],
  onCancel,
  onSubmit,
  instructions,
}) => {
  const { register, handleSubmit, formState, control, setValue } = useForm({
    resolver: yupResolver(schema),
    mode: 'onChange',
  });

  const defaultAttorney = useMemo(
    () =>
      clerkOfTheBoardAttorneys.find(
        (opt) => defaultAttorneyId && opt.value === defaultAttorneyId
      ),
    [defaultAttorneyId, clerkOfTheBoardAttorneys]
  );

  const dispositionOptions = useMemo(() => Object.values(DISPOSITIONS), []);

  const sectionStyle = css({ marginBottom: '24px' });

  useEffect(() => {
    setValue('attorney', defaultAttorney);
  }, [defaultAttorney]);

  return (
    <form
      className="docket-switch-ruling"
      onSubmit={handleSubmit(onSubmit)}
      aria-label="Rule on Docket Switch"
    >
      <AppSegment filledBackground>
        <h1>{sprintf(DOCKET_SWITCH_RULING_TITLE, appellantName)}</h1>
        <div {...sectionStyle}>
          <ReactMarkdown source={DOCKET_SWITCH_RULING_INSTRUCTIONS} />
        </div>
        <hr {...sectionStyle} />
        <div {...sectionStyle}>
          <ReactMarkdown plugins={[gfm]} source={instructions[0]} linkTarget="_blank" />
        </div>
        <hr {...sectionStyle} />

        <RadioField
          name="disposition"
          label="How will you rule on this docket switch?"
          options={dispositionOptions}
          inputRef={register}
          strongLabel
          vertical
        />

        <TextareaField
          inputRef={register}
          name="context"
          label="Provide any additional context and instructions"
          strongLabel
        />

        <div style={{ maxWidth: '35rem' }}>
          <Controller
            as={SearchableDropdown}
            control={control}
            defaultValue={defaultAttorney}
            name="attorney"
            label="Assign to Office of the Clerk of the Board"
            searchable
            options={clerkOfTheBoardAttorneys}
            placeholder="Select attorney"
            strongLabel
          />
        </div>
      </AppSegment>
      <div className="controls cf-app-segment">
        <Button
          type="submit"
          name="submit"
          disabled={!formState.isValid}
          classNames={['cf-right-side']}
        >
          Submit
        </Button>
        {onCancel && (
          <Button
            type="button"
            name="Cancel"
            classNames={['cf-right-side', 'usa-button-secondary']}
            onClick={onCancel}
            styling={{ style: { marginRight: '1rem' } }}
          >
            Cancel
          </Button>
        )}
      </div>
    </form>
  );
};

DocketSwitchRulingForm.propTypes = {
  appellantName: PropTypes.string.isRequired,
  defaultAttorneyId: PropTypes.number,
  clerkOfTheBoardAttorneys: PropTypes.array.isRequired,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
  instructions: PropTypes.array.isRequired
};
