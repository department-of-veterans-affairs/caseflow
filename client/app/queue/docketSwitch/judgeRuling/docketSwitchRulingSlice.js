import { createAsyncThunk } from '@reduxjs/toolkit';
import ApiUtil from '../../../util/ApiUtil';

export const createDocketSwitchGrantedTask = createAsyncThunk(
  'tasks/createDocketSwitchGrantedTask',
  async (data) => {
    try {
      const res = await ApiUtil.post('/tasks', { data });
      const updatedTasks = res.body?.tasks?.data;

      return updatedTasks;
    } catch (error) {
      console.error('Error creating task', error);
      throw error;
    }
  }
);

export const createDocketSwitchDeniedTask = createAsyncThunk(
  'tasks/createDocketSwitchDeniedTask',
  async (data) => {
    try {
      const res = await ApiUtil.post('/tasks', { data });
      const updatedTasks = res.body?.tasks?.data;

      return updatedTasks;
    } catch (error) {
      console.error('Error creating task', error);
      throw error;
    }
  }
);
