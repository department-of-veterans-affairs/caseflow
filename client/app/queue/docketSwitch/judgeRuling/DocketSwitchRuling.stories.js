import React from 'react';

import { DocketSwitchRulingForm } from './DocketSwitchRulingForm';

const clerkOfTheBoardAttorneys = [
  { value: 1, label: 'COB Attorney 1' },
  { value: 2, label: 'COB Attorney 2' },
];

const instructions = ["**Summary:** Summary\n\n**Is this a timely request:** Yes\n\n**Recommendation:** Grant all issues\n\n**Draft letter:** http://www.va.gov"];

export default {
  title: 'Queue/Docket Switch/DocketSwitchRulingForm',
  component: DocketSwitchRulingForm,
  decorators: [],
  parameters: {},
  args: {
    appellantName: 'Jane Doe',
    clerkOfTheBoardAttorneys,
    instructions,
  },
  argTypes: {
    onCancel: { action: 'cancel' },
    onSubmit: { action: 'submit' },
  },
};

const Template = (args) => <DocketSwitchRulingForm {...args} />;

export const Basic = Template.bind({});

Basic.parameters = {
  docs: {
    storyDescription:
      'Used by a judge to make a ruling on a Docket Switch.',
  },
};

export const DefaultAttorney = Template.bind({});
DefaultAttorney.args = {
  defaultAttorneyId: 2
};
