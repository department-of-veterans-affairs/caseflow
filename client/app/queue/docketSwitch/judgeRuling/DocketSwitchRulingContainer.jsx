import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams, useRouteMatch } from 'react-router';
import { taskById, appealWithDetailSelector } from 'app/queue/selectors';
import { taskActionData } from 'app/queue/utils';
import DISPOSITIONS from 'constants/DOCKET_SWITCH_DISPOSITIONS';
import { DocketSwitchRulingForm } from './DocketSwitchRulingForm';
import {
  DOCKET_SWITCH_RULING_SUCCESS_TITLE,
  DOCKET_SWITCH_RULING_SUCCESS_MESSAGE,
} from '../../../../COPY';

import { sprintf } from 'sprintf-js';
import { showSuccessMessage } from '../../uiReducer/uiActions';
import { addressDocketSwitchRuling } from '../docketSwitchSlice';

export const formatDocketSwitchRuling = ({
  disposition,
  hyperlink,
  context,
}) => {
  const parts = [];

  parts.push(`I am proceeding with a ${DISPOSITIONS[disposition].judgeRulingText}.`);
  if (hyperlink) {
    parts.push(`**Signed ruling letter:** [View link](${hyperlink})`);
  }
  parts.push(context);

  return parts.join('  \n  \n');
};

export const DocketSwitchRulingContainer = () => {
  const { appealId, taskId } = useParams();
  const { goBack, push } = useHistory();
  const dispatch = useDispatch();
  const task = useSelector((state) => taskById(state, { taskId }));
  const appeal = useSelector((state) =>
    appealWithDetailSelector(state, { appealId })
  );
  const match = useRouteMatch();
  const { selected, options } = taskActionData({ task, match });

  // Use regex to pull draft letter hyperlink out from the current task's instructions, if possible.
  // It will be reused in the subsequent task's instructions.
  const draftInstructions = task.instructions.join('\n');
  const hyperlink = (/\*\*Draft letter:\*\* \[[^\]]+\]\(([^)]+)\)/).exec(draftInstructions)?.[1];

  // eslint-disable-next-line no-console
  const handleSubmit = async (formData) => {
    const instructions = formatDocketSwitchRuling({ hyperlink, ...formData });
    const { disposition } = formData;
    const dispositionType = DISPOSITIONS[disposition].dispositionType;
    const taskType = `DocketSwitch${dispositionType}Task`;

    const data = {
      task_id: taskId,
      new_task_type: taskType,
      instructions,
      assigned_to_user_id: formData.attorney.value,
    };

    const successMessage = {
      title: sprintf(DOCKET_SWITCH_RULING_SUCCESS_TITLE, dispositionType.toLowerCase(), appeal.appellantFullName),
      detail: DOCKET_SWITCH_RULING_SUCCESS_MESSAGE,
    };

    try {
      await dispatch(addressDocketSwitchRuling(data));
      dispatch(showSuccessMessage(successMessage));
      push('/queue');
    } catch (error) {
      // Perhaps show an alert that indicates error, advise trying again...?
      console.error('Error saving task', error);
    }
  };

  return (
    <DocketSwitchRulingForm
      onCancel={goBack}
      onSubmit={handleSubmit}
      clerkOfTheBoardAttorneys={options}
      defaultAttorneyId={selected}
      appellantName={appeal.appellantFullName}
      instructions={task.instructions}
    />
  );
};
