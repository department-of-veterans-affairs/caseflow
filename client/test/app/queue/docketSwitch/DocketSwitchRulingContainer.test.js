import { formatDocketSwitchRuling } from 'app/queue/docketSwitch/judgeRuling/DocketSwitchRulingContainer';

describe('formatDocketSwitchRuling', () => {
  const context = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit...';

  let disposition = 'granted';

  describe('with granted disposition', () => {
    beforeAll(() => {
      disposition = 'granted';
    });

    it('properly formats', () => {
      const res = formatDocketSwitchRuling({
        context,
        disposition,
      });

      expect(res).toMatch(
        new RegExp('I am proceeding with a full switch\\. {2}\\n {2}\\n')
      );
      expect(res).toMatch(
        new RegExp(context)
      );
      expect(res).toMatchSnapshot();
    });
  });

  describe('with partially_granted disposition', () => {
    beforeAll(() => {
      disposition = 'partially_granted';
    });

    it('properly formats', () => {
      const res = formatDocketSwitchRuling({
        context,
        disposition,
      });


      expect(res).toMatch(
        new RegExp(
          'I am proceeding with a partial switch\\. {2}\\n {2}\\n'
        )
      );
      expect(res).toMatch(
        new RegExp(context)
      );
      expect(res).toMatchSnapshot();
    });
  });

  describe('with denied disposition', () => {
    beforeAll(() => {
      disposition = 'denied';
    });

    it('properly formats', () => {
      const res = formatDocketSwitchRuling({
        context,
        disposition,
      });

      expect(res).toMatch(
        new RegExp('I am proceeding with a denial\\. {2}\\n {2}\\n')
      );
      expect(res).toMatch(
        new RegExp(context)
      );
      expect(res).toMatchSnapshot();
    });
  });
});
