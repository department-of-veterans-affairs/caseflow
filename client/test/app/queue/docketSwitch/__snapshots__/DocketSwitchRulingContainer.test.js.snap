// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatDocketSwitchRuling with denied disposition properly formats 1`] = `
"I am proceeding with a denial.  
  
Lorem ipsum dolor sit amet, consectetur adipiscing elit..."
`;

exports[`formatDocketSwitchRuling with granted disposition properly formats 1`] = `
"I am proceeding with a full switch.  
  
Lorem ipsum dolor sit amet, consectetur adipiscing elit..."
`;

exports[`formatDocketSwitchRuling with partially_granted disposition properly formats 1`] = `
"I am proceeding with a partial switch.  
  
Lorem ipsum dolor sit amet, consectetur adipiscing elit..."
`;
